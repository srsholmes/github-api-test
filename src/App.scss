@import 'stylesheets/colours';

.something {
  color: $mainFont;
  font-size: 2em;
}
