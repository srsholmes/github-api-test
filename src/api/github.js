import axios from 'axios';

const githubGraphQL = axios.create({
  baseURL: 'https://api.github.com/graphql',
  headers: {
    Authorization: `bearer ${process.env.GITHUB_PERSONAL_ACCESS_TOKEN}`
  }
});

const SEARCH_USERS = `
  query ($user: String!) {
    search(query: $user, type: USER, first: 10) {
      edges {
        node {
          ... on User {
            avatarUrl,
            name,
            login
          }
        }
      }
    }
  }
`;

const ACTIVITY = `
  query ($user: String!) {
    repositoryOwner(login: $user) {
      ... on User {
        login
        avatarUrl
        contributionsCollection {
          startedAt
          endedAt
          commitContributionsByRepository(maxRepositories:10) {
            repository {
              name,
            }
            contributions(first:5) {
              edges {
                node {
                  occurredAt
                }
              }
            }
          }
        }
      }
    }
  }
`;

export const searchUser = async val => {
  const res = await githubGraphQL.post('', {
    query: SEARCH_USERS,
    variables: {user: val}
  });
  console.log({res});
  return res.data.data.search.edges;
};

export const getUserActivity = async val => {
  const res = await githubGraphQL.post('', {
    query: ACTIVITY,
    variables: {user: val}
  });
  return res.data.data.repositoryOwner.contributionsCollection
    .commitContributionsByRepository;
};
