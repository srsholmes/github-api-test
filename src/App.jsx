import React, {useState, useEffect} from 'react';
import styles from './App.scss';
import {SearchResults} from './components/SearchResults';
import {SearchInput} from './components/SearchInput';
import {useDebounce} from './hooks/debounce';
import {Loader} from './components/Loader';
import {searchUser} from './api/github';

const App = () => {
  const [searchInput, setSearchInput] = useState('');
  const [results, setResults] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const debouncedSearchTerm = useDebounce(searchInput, 500);

  const setInputAndLoader = e => {
    setSearchInput(e.target.value);
    setIsLoading(true);
  };

  useEffect(() => {
    if (debouncedSearchTerm) {
      searchUser(debouncedSearchTerm).then(x => {
        setResults(x);
        setIsLoading(false);
      });
    } else {
      setIsLoading(false);
      setResults([]);
    }
  }, [debouncedSearchTerm]);

  return (
    <div className={styles.wrapper}>
      <SearchInput setInput={setInputAndLoader} value={searchInput} />
      {isLoading && <Loader />}
      {!isLoading && searchInput.length > 0 && (
        <SearchResults results={results} />
      )}
    </div>
  );
};

export default App;
