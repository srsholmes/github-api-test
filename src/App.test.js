import React from 'react';
import { render, getByTestId, fireEvent } from '@testing-library/react';
import { App } from './App';
import * as dependency from './api/github';

jest.mock('./api/github', () => ({
  searchUser: jest.fn(() => Promise.resolve([]))
}));

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

describe('App', () => {
  it('App', async () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
    const input = getByTestId(container, 'search-input');
    expect(input.value).toBe('');
    fireEvent.change(input, { target: { value: 'testInput' } });
    expect(input.value).toBe('testInput');
    expect(dependency.searchUser).toHaveBeenCalledTimes(0);
    await sleep(600); // Wait for the debounce (500ms)
    expect(dependency.searchUser).toHaveBeenCalledTimes(1);
    expect(dependency.searchUser).toBeCalledWith('testInput');
  });
});
