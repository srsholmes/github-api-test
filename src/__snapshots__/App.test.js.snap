// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App 1`] = `
<div>
  <div
    class="wrapper"
  >
    <h3
      class="title"
    >
      Please search for a Github User:
    </h3>
    <input
      class="input"
      data-testid="search-input"
      type="text"
      value=""
    />
  </div>
</div>
`;
